### PR TITLE
Version 2.10.0 (in progress)

### PR DESCRIPTION
**Pull request by jugglinmike**
We have been hard at work on JSHint version 2.10.0 for over six months now. We're still working, but I'm opening this pull request today in order to give folks a better view into our progress.

Here's a high-level summary of what the next version contains and what we hope to implement before releasing:

- [x] New `esVersion`: 2016 
  - Exponentiation operator
  - Restrictions on the function form of the "use strict" directive
- [ ] New `esVersion`: 2017
  - Async functions
  - Trailing commas in parameter lists
- [x] `unstable` configuration
  - Generic support
  - Options
    - Object "rest" properties
- [x] New linting option: `leanswitch`

If you're interested in adding a new feature to JSHint, we encourage you to work from this branch. However, since we continue to publish bug fixes under the 2.9 release train, **we reserve the right to rebase this branch**. You should also be prepared to rebase your pull request as requested. (This is a more advanced workflow for git, and we'll be happy to assist you as necessary.) Those following along should also note that commits referenced on pages for merged pull requests may be present here under a different SHA.

### CODE REVIEW

**Comment by jugglinmike** [comment]
This branch now supports trailing commas in parameter lists, making `async` functions the final big hurdle

**Comment by jugglinmike** [comment]
Rebased once again. This now includes a re-write of JSHint's class-parsing logic.

Here's the prior version of the branch:

https://github.com/jshint/jshint/tree/v2.10.0-2019-01-13

File: src/lex.js [hunk 1]
```diff
@@ -345,15 +345,22 @@ Lexer.prototype = {
       };
     }
 
-    // 2-character punctuators: <= >= == != ++ -- << >> && ||
-    // += -= *= %= &= |= ^= /=
-    if (ch1 === ch2 && ("+-<>&|".indexOf(ch1) >= 0)) {
+    // 2-character punctuators: ++ -- << >> && || **
+    if (ch1 === ch2 && ("+-<>&|*".indexOf(ch1) >= 0)) {
```
Comment (caitp) [inline src/lex.js]: is this easier to read than `/[+-<>&|*]{2}/.test(ch1 + ch2)` ?

File: src/options.js [hunk 2]
```diff
@@ -1005,6 +1020,41 @@ exports.val = {
...
+   * > Spread properties in object initializers copies own enumerable
+   * > properties from a provided object onto the newly created object.
+   * >
+   * >     let n = { x, y, ...z };
+   * >     n; // { x: 1, y: 2, a: 3, b: 4 }
+   *
+   * https://github.com/tc39/proposal-object-rest-spread
+   */
+  objspreadrest: true
+};
```
Comment (caitp) [inline src/options.js]: I don’t see anything in changes other than tests and error messages where this is actually used. (However, reading this from a phone, so might have missed it)

**Comment by jugglinmike** [comment]
Great news, folks! Following some stellar work by a couple first-time contributors, [I finished up support for async functions](https://github.com/jshint/jshint/pull/3272#issuecomment-458480812). @rwaldron provided excellent review, so everything we set out to do for version 2.10.0 is complete. More than that, actually: we have asynchronous iteration done, too!

You can expect a release on 2019-02-04

**Comment by oskarkrawczyk** [comment]
I'm so happy about the headway made on this and the contribution you've received to!

I'd be lying if I'd say the first thing that I've done upon receiving this comment was not integrate the current code state into JSFiddle, happy to say it works perfectly fine!

Again, such a fantastic job!

**Comment by jugglinmike** [comment]
Thanks so much, Oskar! [JSHint version 2.10.0 is now published.](https://jshint.com/blog/2019-02-05/release-2-10-0/)

**Pull request merged**